docs/code/CrossPlatform.md

```diff
@@ -0,0 +1,175 @@
+ - LdaNative would need to be re-written.
+ - MatrixFactorizationNative would need to be re-written.
+ - MklProxyNative can be ignored. It is not needed with software fallbacks. We will need to modify the build to exclude it as needed.
+ - SymSgdNative would need to be re-written.
+
+### 3.3 Hybrid Between Finding Replacements and Software Fallbacks
+Since some of the native code already has replacements and some of the code already has software fallbacks, we can leverage this work by doing a hybrid between the prior 2 solutions.
+
+At a high level, this solution would require us to:
+ - Fix the build so it's not hardcoded to look for any native dependencies or binaries.
+ - Fix the managed code so that if it can't find the native binaries, the code behaves correctly and performs the software fallback. This includes software fallbacks and/or description error messages.
+ - CpuMathNative mostly has software fallbacks already in place for .NET core 3.1, so only a little work is needed.
+ - FastTreeNative has a flag for a software fallback. We would need to conditionally enable this. Alternatively, we could change the C# code so the software fallback is always enabled for the cases when it can't find the native binaries.
+ - LdaNative builds just find without x86/x64 dependencies, so no replacements would need to be found.
+ - MatrixFactorizationNative builds with the "USESSE" flag that requires x86/x64. We can conditionally enable/disable that flag and no other work would be required.
+ - MklProxyNative is only a wrapper for Intel MKL and can be ignored on non x86/x64 platforms. We will need to modify the build to exclude it as needed.
+ - SymSgdNative needs to be either re-written in managed code, or re-write 4 Intel MKL methods. The 4 methods are just dealing with vector manipulation and shouldn't be hard to do.
+
+## My Suggestion
+My suggestion would be to start with the hybrid approach. It will require the least amount of work to get ML.NET running elsewhere, while still being able to support a large majority of devices out of the gate. This solution will still limit the platforms we can run-on to what we build the native components for, initially Arm64 devices, but we can do a generic Arm64 compile so it should work for all Arm64 v8 devices. The goal is to eventually have a managed implementation which can work everywhere .NET does and accelerated components to increase performance where possible. This could include native components in some cases or hardware intrinsics in others.
+
+### Improving managed fallback through intrinsics
+We should also target .NET 5 so that we gain access to the Arm64 intrinsics. Rather than implementing special-purpose native libraries to take advantage of architecture-specific instructions we should instead enhance performance ensuring our managed implementation leverage intrinsics.
```

Comment:
It may well be OK if Arm* only works on .NET 5+, or it's slow on 3.1 only.

Reply:
I was about to write up something similar in the Support Grid section.

If supporting only .NET Core 5+ saves us a lot of effort or complexity. Perhaps this would be a good time to bump the major version of ML.NET and reduce the number of .NET versions we support?

Reply:
We won't save a lot of effort/complexity just by going to .NET 5. We could gain execution speed if we switched to .NET 5+ and also implemented the intrinsics for it. It may be worth it to update the major version of ML.NET once we have all those hardware intrinsics in place for arm.